Add an option in lod importer to automatically generate missing lods to the model builder component.

rossbridger opened the pull request:
## What does this PR do?

This should add a checkbox on lod importer, which when turned on, will allow the importer to automatically generate the missing lods with meshoptimizer.

## How was this PR tested?

Go to FBX/GLTF settings, add "Level of Detail" modifier, select "Auto Generate Missing Lods" on.

rossbridger commented:
I think meshoptimizer should be sitting in 3rdparty repo though since Meshlet gem also uses it.

moudgils commented on Gems/Atom/RPI/Code/CMakeLists.txt:
```diff
@@ -6,6 +6,15 @@
 #
 #
 
+include(FetchContent)
```
Adding @nick-l-o3de to provide guidance on if this is the correct way to pull in meshoptimizer. Would we instead want to pull in meshoptimer via 3p system? So we use FetchContent within https://github.com/o3de/3p-package-source, build the meshoptimizer 3p package for all platforms and pull that into RPI via that system

nick-l-o3de replied:
I would actually prefer to avoid using the prebuilt 3p package system unless there's a good reason like it takes hours to build or something.

rossbridger replied:
Well, I do think meshoptimizer might be better put to 3p system since Meshlet gem uses it as well (it currently require users to manually compile and add the library tho)

nick-l-o3de commented on Gems/Atom/RPI/Code/CMakeLists.txt:
```diff
@@ -6,6 +6,15 @@
 #
 #
 
+include(FetchContent)
+FetchContent_Declare(
+        meshoptimizer
+        GIT_REPOSITORY "https://github.com/zeux/meshoptimizer.git"
+        GIT_TAG "v0.22"
+)
+set(MESHOPT_INSTALL OFF)
+FetchContent_MakeAvailable(meshoptimizer)
+
```
I would recommend adding some message(STATUS ...) here to indicate hte use of a third party.
I'm working on a similar change for OpenMesh, I write 
```
message(STATUS "WhiteBox gem uses OpenMesh-8.1 (BSD-3-Clause) https://gitlab.vci.rwth-aachen.de:9000/OpenMesh/OpenMesh.git")
message(STATUS "      With patch: ${CMAKE_CURRENT_LIST_DIR}/openmesh-o3de.patch")
```

I also found that the 3rd party library spams libraries into the project at the root, I had to fix it by checking out what targets it made (in your case, just one target, I assume, called `meshoptimizer`?) and do this with it

```cmake

# configure the pulled-in targets - set properties on them for compile and install:
set(OPENMESH_TARGETS OpenMeshCore OpenMeshTools)
foreach(OpenMesh_Target ${OPENMESH_TARGETS})

    # for externally fetched content, we want to pretty much the same compile flags and everything
    # but we can't control code quality, so we want to disable warnings-as-errors just for this scope
    # and then re-enable it after the FetchContent_MakeAvailable call.
    target_compile_options(${OpenMesh_Target} PRIVATE "$<$<COMPILE_LANG_AND_ID:CXX,MSVC>:/WX-$<SEMICOLON>/W0>")
    target_compile_options(${OpenMesh_Target} PRIVATE "$<$<COMPILE_LANG_AND_ID:CXX,Clang,GCC>:-w>")

    # this block makes sure that in IDEs like Visual Studio, they show up in the "External" subfolder under the gem.
    get_property(this_gem_root GLOBAL PROPERTY "@GEMROOT:${gem_name}@")
    ly_get_engine_relative_source_dir(${this_gem_root} relative_this_gem_root)
    set_property(TARGET ${OpenMesh_Target} PROPERTY FOLDER "${relative_this_gem_root}/External")

    # Make aliases 3rdParty::xxxxx for all the OpenMesh targets
    # note that we don't need to call ly_create_alias for OpenMeshCore and OpenMeshTools since they already
    # exist as targets.  ly_create_alias would try to create those base names (without namespace) and then
    # alias the 3rdParty::xxxxx names to them.  Since all that work is alrady done, we only need to make a plain alias.
    add_library(3rdParty::${OpenMesh_Target} ALIAS ${OpenMesh_Target})
endforeach()

```

as an example.

Maybe take a look at the file https://github.com/nick-l-o3de/o3de/blob/move_whitebox_gem_to_use_openmesh_inline/Gems/WhiteBox/Code/BuildOpenMesh.cmake 

(That branch just does `include(BuildOpenMesh.cmake)` in the parent CMakeLists.txt)

Of specific interest is how it makes sure that the library is available in the installer version.  If we don't do that, then people trying to use the installer will be missing the library.

nick-l-o3de reviewed (commented):
We also need to make sure we get the appropriate paper trail for adding a new 3rd party dependency to open 3d engine.  I'm currently investigating what that trail might be.  The license is MIT, so it should be compatible.

rossbridger commented:
Fixed cmake script so that we use 3rdParty::meshoptimizer now. Also added configurable parameters for the lod generation (for some reason I can't figure out how to hide these extra options when lod generation is disabled)

rossbridger commented:
Updated the build system based on the openmesh PR.

moudgils commented on Gems/Atom/RPI/3rdParty/Findmeshoptimizer.cmake:
```diff
@@ -0,0 +1,35 @@
+add_library(3rdParty::${MESHOPTIMIZER_TARGET} ALIAS ${MESHOPTIMIZER_TARGET})
+ly_install(FILES ${CMAKE_CURRENT_LIST_DIR}/Installer/Findmeshoptimizer.cmake DESTINATION cmake/3rdParty)
+
+set(OpenMesh_FOUND TRUE)
```
This needs to be updated. Seems like copy paste error from OpenMesh cmake

moudgils commented on Gems/Atom/RPI/3rdParty/Installer/Findmeshoptimizer.cmake:
```diff
@@ -0,0 +1,22 @@
+add_library(${MESHOPTIMIZER_TARGET} IMPORTED INTERFACE GLOBAL)
+add_library(3rdParty::${MESHOPTIMIZER_TARGET} ALIAS ${MESHOPTIMIZER_TARGET})
+
+set(OpenMesh_FOUND TRUE)
```
This needs to be updated. Seems like copy paste error from OpenMesh cmake

moudgils reviewed (changes requested):
Once the cmake file changes are addressed I will approve.

nick-l-o3de commented:
I think the only remaining checkbox to check here would be, codeowners going forward, and how do we get approval of using a new 3p in the foundation.  Will look into it

KylerBessert commented:
Since meshoptimizer will be added through the atom rpi 3rdparty cmake file, should the meshlets gem be updated to reference this? Right now it states you need to compile meshoptimizer yourself and place it in the designated location to link to, so the meshlets ASV will work. The txt file could specify this change and the meshlets gem could try to utilize meshoptimizer that is grabbed for atom if available.

Not important in this PR, more of a future endeavor

rossbridger commented:
Maybe I can refactor the mesh simplification code into a public available function, like creating a simplified ModelLodAsset from an existing ModelLodAsset.

nick-l-o3de commented:
The TSC is planning to review / vote on whether to include `meshoptimizer` as an approved 3p next meeting.

rossbridger commented:
Refactored the code a bit, also managed to link meshopt library with Meshlet gem (it still can't compile because of the change in Atom API)

rossbridger commented:
Added license file and rebased
